src/main/cpp/Drive/CompFilter.cpp

```diff
@@ -0,0 +1,169 @@
+#include "Drive/CompFilter.h"
+
+#include <algorithm>
+#include <iostream>
+#include <frc/smartdashboard/SmartDashboard.h>
+
+#include "Util/Mathutil.h"
+
+/**
+ * Constructor
+ * 
+ * @param alpha A value between 0 and 1 to determine weight of trust towards wheel odometry
+ * @param maxTime maximum time before discarding measurements, in s
+ * @param posOffset pointer to Position offset in robot cpp
+*/
+CompFilter::CompFilter(double alpha, double maxTime, vec::Vector2D *posOffset) 
+  : m_alpha{std::clamp(alpha, 0.0, 1.0)}, m_maxTime{maxTime}, m_posOffset{posOffset} {
+  m_states[0].pos = *posOffset;
+  m_states[0].vel = {0, 0};
+}
+
+/**
+ * Adds data from wheel odometry
+ * 
+ * @param worldVel Averaged velocities of wheels in world frame (after rotating navX angle)
+ * @param curTime current robot time (from startup), in ms
+*/
+void CompFilter::AddWheelVel(vec::Vector2D worldVel, std::size_t curTime) {
+  auto lastIt = m_states.rbegin(); 
+
+  std::size_t lastTime = lastIt->first;
+  vec::Vector2D lastPos = lastIt->second.pos;
+  double deltaT = (curTime - lastTime) / 1000.0;
+
+  // integrate velocity to get position
+  vec::Vector2D curPos = lastPos + worldVel * deltaT;
+
+  m_states[curTime].pos = curPos;
+  m_states[curTime].vel = worldVel;
+
+  // delete map values > maxTime seconds ago, unless it's the last value in the map
+  auto firstIt = m_states.lower_bound(curTime - static_cast<std::size_t>(m_maxTime * 1000.0)); // first iterator to include
+  m_states.erase(m_states.begin(), firstIt);
+
+  // if empty, insert zeroes
+  // we want to make sure there is at least one value at all times
+  if (m_states.size() == 0) {
+    m_states[curTime].pos = *m_posOffset;
+    m_states[curTime].vel = {0, 0};
+  }
+  // this will mess up odometry but at least no runtime errors
+}
+
+/**
+ * Adds data from camera odometry
+ * 
+ * @param camPos Position of robot given apriltag data
+ * @param timeOffset Difference in time between now and the time that the camera reading was read, in ms
+ * @param curTime current robot time (from startup), in ms
+*/
+void CompFilter::AddCamPos(vec::Vector2D camPos, std::size_t timeOffset, std::size_t curTime) {
+  std::size_t timeOfCam = curTime - timeOffset;
+  auto itOfCam = m_states.lower_bound(timeOfCam);
+
+  // error checking
+  if (itOfCam == m_states.end()) {
+    return;
+  }
+
+  // correct position at time of receiving data from camera
+  // std::size_t timeNearCam = itOfCam->first;
+  vec::Vector2D posNearTime = itOfCam->second.pos;
```

Comment:
Might want to interpolate between the 2 nearest positions, in case cam resolution is increases at the cost of decreasing framerate